### PR TITLE
move some functions to shared

### PR DESCRIPTION
**Pull request by monstermunchkin**
Move the functions `Unpack` and `DownloadFileSha256`, along with some dependencies, to shared.

### CODE REVIEW

File: shared/util_unix.go [hunk 1]
```diff
@@ -13,3 +16,37 @@ func GetOwnerMode(fInfo os.FileInfo) (os.FileMode, int, int) {
 	gid := int(fInfo.Sys().(*syscall.Stat_t).Gid)
 	return mode, uid, gid
 }
+
+func DeviceTotalMemory() (int64, error) {
```
Comment (stgraber) [inline shared/util_unix.go]: That function is Linux-only, /proc/meminfo doesn't exist on other Unix OSes

File: shared/interfaces.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package shared
+
+// The ProgressData struct represents new progress information on an operation
+type ProgressData struct {
```
Comment (stgraber) [inline shared/interfaces.go]: I think this would make more sense in the ioprogress package

File: shared/archive_unix.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+		err1 := syscall.Statfs(path, &fs)
+		if err1 != nil {
+			return err1
+		}
+
+		// Check if we're running out of space
+		if int64(fs.Bfree) < int64(2*fs.Bsize) {
+			if sType == StorageTypeLvm {
+				return fmt.Errorf("Unable to unpack image, run out of disk space (consider increasing your pool's volume.size).")
```
Comment (stgraber) [inline shared/archive_unix.go]: So given that this now applies to more than just LVM and I believe is the reason why you're dragging out the StorageType stuff out of the lxd package (making a huge diff), it's probably easier to just replace the sType argument with a blockBackend (bool) argument and have the warning shown if it's true.

Then in lxd we can call Unpack with blockBackend=true if the backend is LVM or Ceph.

That should significantly reduce the size of this branch.

**Comment by monstermunchkin** [comment]
I addressed the issues.

**Comment by brauner** [comment]
jenkins: test this please

**Pull request merged**